storage/reads/response_writer_test.go

```diff
@@ -123,3 +132,403 @@ func TestResponseWriter_WriteGroupResultSet_Stats(t *testing.T) {
+
+			cur := newSeriesGeneratorSeriesCursor(gen.NewMergedSeriesGenerator(gens))
+			rs := reads.NewFilteredResultSet(context.Background(), &datatypes.ReadFilterRequest{}, cur)
+			err := w.WriteResultSet(rs)
+			if err != nil {
+				t.Fatalf("unexpected err: %v", err)
+			}
+			w.Flush()
+
+			assert.Equal(t, ss, exp)
+		})
+
+		t.Run("multi-series strings", func(t *testing.T) {
+			exp := sendSummary{
+				seriesCount: 4,
+				stringCount: 6900,
+			}
+			var ss sendSummary
+
+			stream := mock.NewResponseStream()
+			stream.SendFunc = ss.makeSendFunc()
+			w := reads.NewResponseWriter(stream, 0)
+
+			var gens []gen.SeriesGenerator
+			gens = append(gens, makeTypedSeries("m0", "t", "s0", strings.Repeat("aaa", 100), 2000, 1))
+			gens = append(gens, makeTypedSeries("m0", "t", "s1", strings.Repeat("bbb", 200), 1500, 1))
+			gens = append(gens, makeTypedSeries("m0", "t", "s2", strings.Repeat("ccc", 300), 2500, 1))
+			gens = append(gens, makeTypedSeries("m0", "t", "s3", strings.Repeat("ddd", 200), 900, 1))
+
+			cur := newSeriesGeneratorSeriesCursor(gen.NewMergedSeriesGenerator(gens))
+			rs := reads.NewFilteredResultSet(context.Background(), &datatypes.ReadFilterRequest{}, cur)
+			err := w.WriteResultSet(rs)
+			if err != nil {
+				t.Fatalf("unexpected err: %v", err)
+			}
+			w.Flush()
+
+			assert.Equal(t, ss, exp)
+		})
+
+		t.Run("writer doesn't send series with no values", func(t *testing.T) {
+			exp := sendSummary{
+				seriesCount: 2,
+				stringCount: 3700,
+			}
+			var ss sendSummary
+
+			stream := mock.NewResponseStream()
+			stream.SendFunc = ss.makeSendFunc()
+			w := reads.NewResponseWriter(stream, 0)
+
+			var gens []gen.SeriesGenerator
+			gens = append(gens, makeTypedSeries("m0", "t", "s0", strings.Repeat("aaa", 100), 2000, 1))
+			gens = append(gens, makeTypedSeries("m0", "t", "s1", strings.Repeat("bbb", 200), 0, 1))
+			gens = append(gens, makeTypedSeries("m0", "t", "s2", strings.Repeat("ccc", 100), 1700, 1))
+			cur := newSeriesGeneratorSeriesCursor(gen.NewMergedSeriesGenerator(gens))
+
+			rs := reads.NewFilteredResultSet(context.Background(), &datatypes.ReadFilterRequest{}, cur)
+			err := w.WriteResultSet(rs)
+			if err != nil {
+				t.Fatalf("unexpected err: %v", err)
+			}
+			w.Flush()
+
+			assert.Equal(t, ss, exp)
+		})
+	})
+
+	t.Run("error conditions", func(t *testing.T) {
+		t.Run("writer returns stream error", func(t *testing.T) {
+			exp := errors.New("no write")
+
+			stream := mock.NewResponseStream()
+			stream.SendFunc = func(r *datatypes.ReadResponse) error { return exp }
+			w := reads.NewResponseWriter(stream, 0)
+
+			cur := newSeriesGeneratorSeriesCursor(makeTypedSeries("m0", "t", "f0", strings.Repeat("0", 1000), 2000, 1))
+			rs := reads.NewFilteredResultSet(context.Background(), &datatypes.ReadFilterRequest{}, cur)
+			_ = w.WriteResultSet(rs)
+			assert.Equal(t, w.Err(), exp)
+		})
+	})
+
+	t.Run("issues", func(t *testing.T) {
+		t.Run("#4321 short write", func(t *testing.T) {
```

Comment:
Maybe this one should be called just "short write", since we'll forget about issue #4321 in a few days.

Reply:
Yes, I will remove the issue reference. I had seen it in a couple of other places, but I don’t think it matters

Reply:
Done